fe1-web/src/features/digital-cash/objects/transaction/Transaction.ts

```diff
@@ -0,0 +1,388 @@
+import { Base64UrlData, Hash, KeyPair, PopToken, PublicKey } from 'core/objects';
+import { SCRIPT_TYPE, COINBASE_HASH } from 'resources/const';
+
+import { isDefined } from '../../../../core/types';
+import { TransactionInput, TransactionInputJSON, TransactionInputState } from './TransactionInput';
+import {
+  TransactionOutput,
+  TransactionOutputJSON,
+  TransactionOutputState,
+} from './TransactionOutput';
+
+export interface TransactionJSON {
+  version: number;
+  inputs: TransactionInputJSON[];
+  outputs: TransactionOutputJSON[];
+  lock_time: number;
+}
+
+export interface TransactionState {
+  version: number;
+  inputs: TransactionInputState[];
+  outputs: TransactionOutputState[];
+  lockTime: number;
+  transactionId?: string;
+}
+
+/**
+ * A coin transaction object
+ */
+export class Transaction {
+  public readonly version: number;
+
+  public readonly inputs: TransactionInput[];
+
+  public readonly outputs: TransactionOutput[];
+
+  public readonly lockTime: number;
+
+  public readonly transactionId: Hash;
+
+  constructor(obj: Partial<Transaction>) {
+    if (obj === undefined || obj === null) {
+      throw new Error(
+        'Error encountered while creating a Transaction object: undefined/null parameters',
+      );
+    }
+    if (obj.version === undefined) {
+      throw new Error("Undefined 'version' when creating 'Transaction'");
+    }
+    if (obj.inputs === undefined) {
+      throw new Error("Undefined 'inputs' when creating 'Transaction'");
+    }
+    if (obj.inputs.length === 0) {
+      throw new Error("Empty 'inputs' when creating 'Transaction'");
+    }
+    if (obj.outputs === undefined) {
+      throw new Error("Undefined 'outputs' when creating 'Transaction'");
+    }
+    if (obj.outputs.length === 0) {
+      throw new Error("Empty 'outputs' when creating 'Transaction'");
+    }
+    if (obj.lockTime === undefined) {
+      throw new Error("Undefined 'lockTime' when creating 'Transaction'");
+    }
+
+    this.version = obj.version;
+    this.inputs = obj.inputs;
+    this.outputs = obj.outputs;
+    this.lockTime = obj.lockTime;
+
+    if (obj.transactionId === undefined) {
+      this.transactionId = this.hashTransaction();
+    } else {
+      if (obj.transactionId.valueOf() !== this.hashTransaction().valueOf()) {
+        throw new Error(
+          "The computed transaction hash does not correspond to the provided one when creating 'Transaction'",
+        );
+      }
+      this.transactionId = obj.transactionId;
+    }
+  }
+
+  /**
+   * Hashes the transaction to get its id
+   */
```

Comment:
In general I think it would be great to link the specification for complex things like this but should be okay for this :)